app/src/main/java/com/smile/smilesms/SmsHandler/Message.java

```
@@ -0,0 +1,85 @@
+package com.smile.smilesms.SmsHandler;
+
+import com.smile.smilesms.SmsHandler.exceptions.InvalidHeaderException;
+import com.smile.smilesms.SmsHandler.exceptions.InvalidPayloadException;
+
+/**
+ * DataUnit abstract class
+ * @author Giovanni Barca
+ *
+ * @param <H> Header data-type
+ * @param <M> Payload data-type
+ */
+public abstract class Message<H extends Header, M extends Payload> {
+    protected H header;
+    protected M payload;
+
+    /**
+     * Create a message
+     * @param header Header of the DataUnit
+     * @param payload Payload of the DataUnit
+     * @throws InvalidHeaderException when an invalid header is passed
+     * @throws InvalidPayloadException when an invalid payload is passed
+     */
+    public Message(H header, M payload) throws InvalidHeaderException, InvalidPayloadException {
+        if (header == null || !header.isValid())
+            throw new InvalidHeaderException();
+
+        if (payload == null || !payload.isValid())
+            throw new InvalidPayloadException();
+
+        this.payload = payload;
+        this.header = header;
+    }
+
+    /**
+     * Get header of the DataUnit
+     * @return header of the DataUnit
+     */
+    public H getHeader() {
+        return header;
+    }
+
+    /**
+     * Get payload of the DataUnit
+     * @return the payload of the DataUnit
+     */
+    public M getPayload() {
+        return payload;
+    }
+
+    /**
+     * Set header of the DataUnit
+     * @param header Header to assign to the DataUnit
+     * @return true if header is valid, false if header is not valid
+     */
+    public boolean setHeader(H header){
+        if (header == null || !header.isValid())
+            return false;
```

Instead of returning a boolean value consider throwing an `InvalidHeaderException`, like you do in the constructor.